src/gt4py/next/iterator/transforms/infer_domain.py

```diff
@@ -0,0 +1,206 @@
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either version 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+import dataclasses
+
+from gt4py.eve.extended_typing import Dict, Tuple, Union
+from gt4py.next.common import Dimension
+from gt4py.next.iterator import ir as itir
+from gt4py.next.iterator.ir_utils import ir_makers as im
+from gt4py.next.iterator.transforms.global_tmps import SymbolicDomain, SymbolicRange, domain_union
+from gt4py.next.iterator.transforms.trace_shifts import TraceShifts
+
+
+@dataclasses.dataclass(frozen=True)
+class InferDomain:
+    @staticmethod
+    def _get_symbolic_domain(domain: Union[SymbolicDomain, itir.FunCall]) -> SymbolicDomain:
+        if isinstance(domain, SymbolicDomain):
+            return domain
+        if isinstance(domain, itir.FunCall) and domain.fun == im.ref("cartesian_domain"):
+            return SymbolicDomain.from_expr(domain)
+        raise TypeError("domain must either be a FunCall or a SymbolicDomain.")
+
+    @staticmethod
+    def _merge_domains(
+        original_domains: Dict[str, SymbolicDomain], new_domains: Dict[str, SymbolicDomain]
+    ) -> Dict[str, SymbolicDomain]:
+        for key, value in new_domains.items():
+            if key in original_domains:
+                original_domains[key] = domain_union([original_domains[key], value])
+            else:
+                original_domains[key] = value
+
+        return {
+            key: domain_union(value) if isinstance(value, list) else value
+            for key, value in original_domains.items()
+        }
+
+    @classmethod
+    def _translate_domain(
+        cls,
+        symbolic_domain: SymbolicDomain,
+        shift: Tuple[itir.OffsetLiteral, itir.OffsetLiteral],
+        offset_provider: Dict[str, Dimension],
+    ) -> SymbolicDomain:
+        dims = list(symbolic_domain.ranges.keys())
+        new_ranges = {dim: symbolic_domain.ranges[dim] for dim in dims}
+        if shift:
+            off, val = shift
+            current_dim = offset_provider[off.value]
+            new_ranges[current_dim] = SymbolicRange.translate(
+                symbolic_domain.ranges[current_dim], val.value
+            )
+        return SymbolicDomain("cartesian_domain", new_ranges)
+
+    @classmethod
+    def infer_as_fieldop(
+        cls,
+        applied_fieldop: itir.FunCall,
+        input_domain: SymbolicDomain,
+        offset_provider: Dict[str, Dimension],
+    ) -> Tuple[itir.FunCall, Dict[str, SymbolicDomain]]:  # todo: test scan operator
+        assert isinstance(applied_fieldop, itir.FunCall) and isinstance(
```

Comment:
```suggestion
        if isinstance(applied_fieldop, itir.SymRef):
          accessed_domains = {applied_fieldop.id: input_domain}
          return applied_fieldop, accessed_domains
          
        assert isinstance(applied_fieldop, itir.FunCall) and isinstance(
```
This should work for your example @havogt here https://github.com/GridTools/gt4py/pull/1568/files#r1672562626. Tuples and lets are not yet supported by this pass. The latter can be solved by inlining first.

Reply:
Implemented (in the let branch).